### PR TITLE
[Zoning Project] Entire Code Base Refactor

### PR DESCRIPTION
**Pull request by nanjiangwill**
Design Idea
- Separate the process of extracting information from pdf, giving information to LLM and analyzing the result from LLM
  - So that in the future it is easier to do sweep over different configurations with different components
- Using wandb + hydra to record the config and to visualize the intermediate result

Todo
- [ ] logging/warning system
- [ ] Type checking 
- [ ] utils.py is too huge, need to find a concise way 
- [ ] wandb
- [ ] Minor change/debug
- [ ] migrating other prompting/agent methods

Check
- [ ] should we publish dataset to HF?
- [ ] use jsonlines over json?

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+-   repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v2.3.0
+    hooks:
+    -   id: check-yaml
+    -   id: end-of-file-fixer
+    -   id: trailing-whitespace
+-   repo: https://github.com/psf/black
+    rev: 22.10.0
+    hooks:
+    -   id: black
```
Comment (srush) [inline .pre-commit-config.yaml:17]: Let's add mypy as well for typing.

File: config/llm/base.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# @package _global_
```
Comment (srush) [inline config/llm/base.yaml]: I think maybe just one global config file? this is a bit too fine-grained.

File: zoning/extract/extractor/textract.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+import os
+import pandas as pd
+import pyarrow as pa
+from tqdm.contrib.concurrent import process_map
+from datasets import load_dataset, Dataset, DatasetDict
+import tqdm
+import numpy as np
+import openai
+
+SCHEMA = pa.schema(
```
Comment (srush) [inline zoning/extract/extractor/textract.py]: Do we need to do this in PyArrow or can it just be a dataclass?

File: zoning/extract/extractor/textract.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+import time
+import json
+import os
+import pandas as pd
+import pyarrow as pa
+from tqdm.contrib.concurrent import process_map
+from datasets import load_dataset, Dataset, DatasetDict
+import tqdm
+import numpy as np
+import openai
```
Comment (srush) [inline zoning/extract/extractor/textract.py]: don't think we need openAI here.

File: zoning/extract/extractor/textract.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+        Checks whether document analysis still in progress
+        """
+        status: str = "IN_PROGRESS"
+        while status == "IN_PROGRESS":
+            time.sleep(5)
+            response = self.extractor.get_document_analysis(JobId=job_id)
+            status = response["JobStatus"]
+            yield status, response.get("StatusMessage", None)
+
+    def get_job_results(self, job_id: str):
```
Comment (srush) [inline zoning/extract/extractor/textract.py]: let's add return types to these

File: zoning/extract/extractor/textract.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
+from base import *
```
Comment (srush) [inline zoning/extract/extractor/textract.py]: This file is too deep. Let's keep everytyhing basically in zoning/

File: zoning/extract/extractor/textract.py (hunk 3 above)
Comment (nanjiangwill) [reply zoning/extract/extractor/textract.py]: I deleted it in the latest commit

File: zoning/extract/extractor/textract.py (hunk 4 above)
Comment (nanjiangwill) [reply zoning/extract/extractor/textract.py]: Deleted

**Pull request merged**